File: src/Interview_exp/contests/bi_weekly/27/winner/CourseScheduleII.cpp
/*LC1462: Course Schedule IV
https://leetcode.com/problems/course-schedule-iv/
There are a total of n courses you
have to take, labeled from 0 to n-1.
Some courses may have direct prerequisites,
for example, to take course 0 you have
first to take course 1, which is expressed as a pair: [1,0]

Given the total number of courses n, a
list of direct prerequisite pairs and
a list of queries pairs.

You should answer for each queries[i] whether
the course queries[i][0] is a prerequisite
of the course queries[i][1] or not.

Return a list of boolean, the answers to the
given queries.

Please note that if course a is a prerequisite
of course b and course b is a prerequisite of
course c, then, course a is a prerequisite of course c.

Example 1:
Input: n = 2, prerequisites = [[1,0]], queries = [[0,1],[1,0]]
Output: [false,true]
Explanation: course 0 is not a prerequisite of course 1 but the opposite is true.

Example 2:
Input: n = 2, prerequisites = [], queries = [[1,0],[0,1]]
Output: [false,false]
Explanation: There are no prerequisites and each course is independent.

Example 3:
Input: n = 3, prerequisites = [[1,2],[1,0],[2,0]], queries = [[1,0],[1,2]]
Output: [true,true]

Example 4:
Input: n = 3, prerequisites = [[1,0],[2,0]], queries = [[0,1],[2,0]]
Output: [false,true]

Example 5:
Input: n = 5, prerequisites = [[0,1],[1,2],[2,3],[3,4]], queries = [[0,4],[4,0],[1,3],[3,0]]
Output: [true,false,true,false]*/
#include <algorithm>
#include <bitset>
#include <cassert>
#include <chrono>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <map>
#include <queue>
#include <random>
#include <set>
#include <stack>
#include <vector>

using namespace std;
// BEGIN NO SAD
#define all(x) x.begin(), x.end()
#define sz(x) (int)(x).size()
typedef vector<int> vi;
// END NO SAD
// REMEMBER CLEAR GLOBAL STATE
bool can[105][105];
class Solution {
public:
    vector<bool> checkIfPrerequisite(int n, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
      for(int i = 0; i < n; i++) for(int j = 0; j < n; j++) can[i][j] = false;
      for(int i = 0; i < n; i++) can[i][i] = true;
      for(auto out: prerequisites) {
        can[out[0]][out[1]] = true;
      }
      for(int i = 0; i < n; i++) for(int j = 0; j < n; j++) for(int k = 0; k < n; k++) can[i][j] |= can[i][k] && can[k][j];
      vector<bool> ret;
      for(auto out: queries) ret.push_back(can[out[0]][out[1]]);
      return ret;
    }
};

File: src/Interview_exp/contests/bi_weekly/27/CourseScheduleII.java
/*LC210. Course Schedule II
https://leetcode.com/problems/course-schedule-ii/
There are a total of n courses you have to take,
* labeled from 0 to n-1.

Some courses may have prerequisites, for example
* to take course 0 you have to first take course 1,
* which is expressed as a pair: [0,1]

Given the total number of courses and a list of
* prerequisite pairs, return the ordering of courses
* you should take to finish all courses.

There may be multiple correct orders, you just need to
* return one of them. If it is impossible to finish all courses, return an empty array.

Example 1:
Input: 2, [[1,0]]
Output: [0,1]
Explanation: There are a total of 2 courses to take. To take course 1 you should have finished
             course 0. So the correct course order is [0,1] .
Example 2:
Input: 4, [[1,0],[2,0],[3,1],[3,2]]
Output: [0,1,2,3] or [0,2,1,3]
Explanation: There are a total of 4 courses to take. To take course 3 you should have finished both
             courses 1 and 2. Both courses 1 and 2 should be taken after you finished course 0.
             So one correct course order is [0,1,2,3]. Another correct ordering is [0,2,1,3] .
Note:*/
package data_structure.graph.dfs;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
class Sln1 {
    List<Integer>[] g;
    public static void main(String[] args) {
        CourseScheduleII courseScheduleII = new CourseScheduleII();
        courseScheduleII.findOrder(4, new int[][]{{1, 0}, {2, 0}, {3, 1}, {3, 2}});
    }
    void buildGraph(int n,int[][] pre){
        for(int i=0;i<n;i++){
            g[i]=new ArrayList<>();
        }
        for(int[] e:pre){
            g[e[1]].add(e[0]);
        }
    }
    void dfs(int i,boolean[] visited, Deque<Integer> stack){
        if(visited[i])
            return;
        visited[i]=true;
        List<Integer> nbrs=g[i];
        for(int nbr:nbrs){
            if(!visited[nbr])
                dfs(nbr, vis, stack);
        }
        stack.add(i);
    }
    public int[] findOrder(int n,int[][] pre){
        if(n<=1)
            return new int[]{};
        g=new ArrayList[n];
        buildGraph(n,pre);
        boolean[] visited=new boolean[n];
        Deque<Integer> stack=new ArrayDeque<>();
        for(int i=0;i<n;i++){
            if(!visited[i])
                dfs(i,visited,stack);
        }
        return stack.toArray();
    }
}

class Sln2 {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        if (numCourses == 0) return null;
        // Convert graph presentation from edges to indegree of adjacent list.
        int indegree[] = new int[numCourses], order[] = new int[numCourses], index = 0;
        for (int i = 0; i < prerequisites.length; i++) // Indegree - how many prerequisites are needed.
            indegree[prerequisites[i][0]]++;

        Queue<Integer> queue = new LinkedList<Integer>();
        for (int i = 0; i < numCourses; i++)
            if (indegree[i] == 0) {
                // Add the course to the order because it has no prerequisites.
                order[index++] = i;
                queue.offer(i);
            }

        // How many courses don't need prerequisites.
        while (!queue.isEmpty()) {
            int prerequisite = queue.poll(); // Already finished this prerequisite course.
            for (int i = 0; i < prerequisites.length; i++) {
                if (prerequisites[i][1] == prerequisite) {
                    indegree[prerequisites[i][0]]--;
                    if (indegree[prerequisites[i][0]] == 0) {
                        // If indegree is zero, then add the course to the order.
                        order[index++] = prerequisites[i][0];
                        queue.offer(prerequisites[i][0]);
                    }
                }
            }
        }
        return (index == numCourses) ? order : new int[0];
    }
}